### PR TITLE
=doc #17655 improvements in streams quickstart

### LINKED ISSUE #17655: Stream Scala Documentation issues for 1.2 Quick Start Guide
1. Task description (“In our case let’s say we want to find all twitter handlers”) is too far from the beginning of chapter.
2. "Both Sources and Flows provide stream operations that can be used to transform the flowing data, a Sink however does not since its the “end of stream” - in following chapter there is Sink.fold() - per my understanding it’s transforming data.
3. For 1.2.1 need to make snippet of code that we can copy and paste into REPL and run. This will simplify understanding and will provide general vision.
4. It’s good to have images for:
  1. flattening sequence (1.2.2)
  2. broadcasting stream (1.2.3)
  3. fun-in and fun-out
5. In code snippet for 1.2.3 it’s not clear what FlowGraph.closed() is and there is no comment on this
6. 1.2.5 introducing term FoldSink - it’s confusing. There is no such class or entity in API, value name is sumSink - consequently no exact matches
7. “Remember those mysterious type parameters” 1.2.6 - no coma between Source and Flow. Also need to add generics them as well. 
8. In Quick start chapter there is no example for Flow. We can see Sink, Source and Flow Graph
9. Would be good to have Glossary with terms and descriptions
10. In 1.2.5 there is mention of Keep.right - which is logical but confusing part that it’s not clear what left is doing and why it’s important to have both.

### PR DESCRIPTION
**Pull request by ktoso**
Resolves #17655
except better explaining `Keep`, I'll think of something later.

### CODE REVIEW

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 1]
```diff
@@ -32,19 +41,17 @@ Let's assume we have a stream of tweets readily available, in Akka this is expre
...
-finally be consumed by a :class:`Sink<In>`. The first type
-parameter—:class:`Tweet` in this case—designates the kind of elements produced
-by the source while the second one describes the object that is created during
-materialization (see below)—:class:`BoxedUnit` (from the ``scala.runtime``
-package) means that no value is produced, it is the generic equivalent of
-``void``. Both Sources and Flows provide stream operations that can be used to
-transform the flowing data, a :class:`Sink` however does not since its the "end
-of stream" and its behavior depends on the type of :class:`Sink` used.
+Streams always start flowing from a ``Source<Out,M1>`` then can continue through ``Flow<In,Out,M2>`` elements or
+more advanced graph elements to finally be consumed by a ``Sink[In,M3>``.
```
Comment (drewhk) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: remaining bracket

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 2]
```diff
@@ -94,7 +105,14 @@ input port to all of its output ports.
 
 Akka Streams intentionally separate the linear stream structures (Flows) from the non-linear, branching ones (FlowGraphs)
 in order to offer the most convenient API for both of these cases. Graphs can express arbitrarily complex stream setups
-at the expense of not reading as familiarly as collection transformations. It is also possible to wrap complex computation
+at the expense of not reading as familiarly as collection transformations. 
+
+
+A graph can be either ``closed`` which is also known as a "*fully connected graph*", or ``partial`` which can be seen as
```
Comment (drewhk) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: maybe link to my page later

**Comment by drewhk** [comment]
LGTM

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 2 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: yeah, gladly - merge yours first when done and I'll amend here

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 1 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: thanks

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 3]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
+Remember those mysterious ``Mat`` type parameters on ``Source<+Out, +Mat>``, ``Flow<-In, +Out, +Mat>`` and ``Sink<-In, +Mat>``?
+They represent the type of values these processing parts return when materialized. When you chain these together,
+you can explicitly combine their materialized values: in our example we used the ``Keep.right`` predefined function,
+which tells the implementation to only care about the materialized type of the stage currently appended to the right.
+As you can notice, the materialized type of sumSink is ``Future<Int>`` and because of using ``Keep.right``,
+the resulting :class:`RunnableGraph` has also a type parameter of ``Future<Int>``.
+
+This step does *not* yet materialize the
+processing pipeline, it merely prepares the description of the Flow, which is now connected to a Sink, and therefore can
+be ``run()``, as indicated by its type: ``RunnableGraph<Future<Int>>``. Next we call ``run()`` which uses the implicit :class:`ActorMaterializer`
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: Remove "implicit" from here.

File: akka-docs-dev/rst/java/stream-flows-and-basics.rst [hunk 4]
```diff
@@ -4,6 +4,8 @@
 Basics and working with Flows
 #############################
 
+.. _core-concepts-scala:
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-flows-and-basics.rst]: This header should be "java"

File: akka-docs-dev/rst/java/stream-flows-and-basics.rst [hunk 5]
```diff
@@ -41,6 +43,7 @@ will use asynchronous means to slow down a fast producer, without blocking its t
 design, since entities that need to wait (a fast producer waiting on a slow consumer) will not block the thread but
 can hand it back for further use to an underlying thread-pool.
 
+.. _defining-and-running-streams-scala:
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-flows-and-basics.rst]: This should be "java" as well.

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 6]
```diff
@@ -56,13 +63,17 @@ For now let's simply print each author:
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: The paragraph above is lacking materialized value type parameter on Sink and Source types. We should have add it at least for consistency.

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 7]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
 this question in a streaming setting would to create a stream of counts described as "*up until now*, we've processed N tweets"),
 but in general it is possible to deal with finite streams and come up with a nice result such as a total count of elements.
 
-First, let's write such an element counter using :class:`FoldSink` and then we'll see how it is possible to obtain materialized
-values from a :class:`MaterializedMap` which is returned by materializing an Akka stream. We'll split execution into multiple
-lines for the sake of explaining the concepts of ``Materializable`` elements and ``MaterializedType``
+First, let's write such an element counter using `Flow.of(Class)`` and ``Sink.fold`` to see how the types look like:
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: One back tick missing on Flow.

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 8]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
 
-First, we prepare the :class:`FoldSink` which will be used to sum all ``Integer`` elements of the stream.
-Next we connect the ``tweets`` stream though a ``map`` step which converts each tweet into the number ``1``,
-finally we connect the flow ``to`` the previously prepared Sink. Notice that this step does *not* yet materialize the
-processing pipeline, it merely prepares the description of the Flow, which is now connected to a Sink, and therefore can
-be ``run()``, as indicated by its type: :class:`RunnableGraph`. Next we call ``run()`` which uses the implicit :class:`ActorMaterializer`
-to materialize and run the flow. The value returned by calling ``run()`` on a ``RunnableGraph`` or ``FlowGraph`` is ``MaterializedMap``,
-which can be used to retrieve materialized values from the running stream.
+First we prepare a reusable ``Flow`` that will change each incoming tweet into an integer of value ``1``.
+We'll use this in order to combine those ones with a ``Sink.fold`` will sum all ``Int`` elements of the stream
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: We combine all values of transformed stream using a ``Sink.fold`` that sums all ...

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 9]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
+and make its result available as a ``Future<Int>``. Next we connect the ``tweets`` stream though a ``map`` step which
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: This mentions a flow that converts tweets to numbers the second time. And looking at the actual doc code this should stay, and the first mention should go.

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 10]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
+converts each tweet into the number ``1``, finally we connect the flow using ``toMat`` the previously prepared Sink.
 
-In order to extract an materialized value from a running stream it is possible to call ``get(Materializable)`` on a materialized map
-obtained from materializing a flow or graph. Since ``FoldSink`` implements ``Materializable`` and implements the ``MaterializedType``
-as ``Future<Integer>`` we can use it to obtain the :class:`Future` which when completed will contain the total length of our tweets stream.
+Remember those mysterious ``Mat`` type parameters on ``Source<+Out, +Mat>``, ``Flow<-In, +Out, +Mat>`` and ``Sink<-In, +Mat>``?
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: Remove + and - variance annotations from java text.

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 3 above)
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: s/Int/Integer

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 11]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
+you can explicitly combine their materialized values: in our example we used the ``Keep.right`` predefined function,
+which tells the implementation to only care about the materialized type of the stage currently appended to the right.
+As you can notice, the materialized type of sumSink is ``Future<Int>`` and because of using ``Keep.right``,
+the resulting :class:`RunnableGraph` has also a type parameter of ``Future<Int>``.
+
+This step does *not* yet materialize the
+processing pipeline, it merely prepares the description of the Flow, which is now connected to a Sink, and therefore can
+be ``run()``, as indicated by its type: ``RunnableGraph<Future<Int>>``. Next we call ``run()`` which uses the implicit :class:`ActorMaterializer`
+to materialize and run the flow. The value returned by calling ``run()`` on a ``RunnableGraph<T>`` is of type ``T``.
+In our case this type is ``Future<Int>`` which, when completed, will contain the total length of our tweets stream.
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: s/Int/Integer

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 12]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
+converts each tweet into the number ``1``, finally we connect the flow using ``toMat`` the previously prepared Sink.
 
-In order to extract an materialized value from a running stream it is possible to call ``get(Materializable)`` on a materialized map
-obtained from materializing a flow or graph. Since ``FoldSink`` implements ``Materializable`` and implements the ``MaterializedType``
-as ``Future<Integer>`` we can use it to obtain the :class:`Future` which when completed will contain the total length of our tweets stream.
+Remember those mysterious ``Mat`` type parameters on ``Source<+Out, +Mat>``, ``Flow<-In, +Out, +Mat>`` and ``Sink<-In, +Mat>``?
+They represent the type of values these processing parts return when materialized. When you chain these together,
+you can explicitly combine their materialized values: in our example we used the ``Keep.right`` predefined function,
+which tells the implementation to only care about the materialized type of the stage currently appended to the right.
+As you can notice, the materialized type of sumSink is ``Future<Int>`` and because of using ``Keep.right``,
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: s/Int/Integer

File: akka-docs-dev/rst/java/stream-quickstart.rst [hunk 13]
```diff
@@ -136,26 +156,30 @@ While this question is not as obvious to give an answer to in case of an infinit
...
+the resulting :class:`RunnableGraph` has also a type parameter of ``Future<Int>``.
```
Comment (2m) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: s/Int/Integer

**Comment by 2m** [comment]
LGTM after some minor nitpicks for the java side of docs.

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 3 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: thanks

File: akka-docs-dev/rst/java/stream-flows-and-basics.rst (hunk 4 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-flows-and-basics.rst]: thanks

File: akka-docs-dev/rst/java/stream-flows-and-basics.rst (hunk 5 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-flows-and-basics.rst]: thanks

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 8 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: nice

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 10 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: thanks

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 11 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: thanks

File: akka-docs-dev/rst/java/stream-quickstart.rst (hunk 6 above)
Comment (ktoso) [inline akka-docs-dev/rst/java/stream-quickstart.rst]: thanks done

**Pull request merged**